utils/signer/src/lib.rs

```diff
@@ -0,0 +1,194 @@
+pub async fn sign_transaction_request_inner(
+    signer: Signer,
+    l1_rpc: Url,
+    mut transaction_request: TransactionRequest,
+) -> Result<TransactionReceipt> {
+    match signer {
+        Signer::Web3Signer(signer_url, signer_address) => {
+            // Set the from address to the signer address.
+            transaction_request.set_from(signer_address);
+
+            // Fill the transaction request with all of the relevant gas and nonce information.
+            let provider = ProviderBuilder::new().network::<Ethereum>().connect_http(l1_rpc);
+            let filled_tx = provider.fill(transaction_request).await?;
+
+            // Sign the transaction request using the Web3Signer.
+            let web3_provider =
+                ProviderBuilder::new().network::<Ethereum>().connect_http(signer_url);
+            let signer = Web3Signer::new(web3_provider.clone(), signer_address);
+
+            let tx = filled_tx.as_builder().unwrap().clone();
+
+            // NOTE: This is a hack because there is not a "data" field on the TransactionRequest.
+            // `eth_signTransaction` expects a "data" field with the calldata.
+            // TODO: Once alloy fixes this, we can remove this wrapper.
+            let wrapper = TransactionRequestWrapper::new(tx);
+
+            let raw: Bytes =
+                signer.provider().client().request("eth_signTransaction", (wrapper,)).await?;
+
+            let tx_envelope = TxEnvelope::decode_2718(&mut raw.as_ref()).unwrap();
+
+            let receipt = provider
+                .send_tx_envelope(tx_envelope)
+                .await
+                .context("Failed to send transaction")?
+                .with_required_confirmations(NUM_CONFIRMATIONS)
+                .with_timeout(Some(Duration::from_secs(TIMEOUT_SECONDS)))
+                .get_receipt()
+                .await?;
+
+            Ok(receipt)
+        }
+        Signer::LocalSigner(private_key) => {
+            let provider = ProviderBuilder::new()
+                .network::<Ethereum>()
+                .wallet(EthereumWallet::new(private_key.clone()))
+                .connect_http(l1_rpc);
+
+            // Set the from address to the Ethereum wallet address.
+            transaction_request.set_from(private_key.address());
+
+            // Fill the transaction request with all of the relevant gas and nonce information.
+            let filled_tx = provider.fill(transaction_request).await?;
+
+            let receipt = provider
+                .send_tx_envelope(filled_tx.as_envelope().unwrap().clone())
+                .await
+                .context("Failed to send transaction")?
+                .with_required_confirmations(NUM_CONFIRMATIONS)
+                .with_timeout(Some(Duration::from_secs(TIMEOUT_SECONDS)))
+                .get_receipt()
+                .await?;
+
+            Ok(receipt)
+        }
+    }
+}
+
+/// A wrapper around `TransactionRequest` that adds a data field as bytes.
```

Comment:
This comment is a bit confusing. Is it just that TransactionRequest doesn't have a data field, but web3signer expects it? That makes sense, but it doesn't seem like a bug in alloy? It's just not what TransactionRequest is for